handshake/src/types/tx.rs

```diff
@@ -0,0 +1,766 @@
+    ///
+    /// For BIP143 (Witness and Compatibility sighash) documentation, see here:
+    ///
+    /// - https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki
+    ///
+    /// TODO: memoize
+    fn hash_sequence(&self, sighash_flag: Sighash) -> TxResult<Blake2b256Digest> {
+        let flag = sighash_flag as u8;
+        if (flag & Sighash::ACP as u8) == Sighash::ACP as u8
+                || (flag & 0x1f) == Sighash::Single as u8
+                || (flag & 0x1f) == Sighash::SingleReverse as u8
+                || (flag & 0x1f) == Sighash::None as u8 {
+            Ok(Blake2b256Digest::default())
+        } else {
+            let mut w = Blake2b256Writer::default();
+            for input in self.vin.iter() {
+                Self::write_u32_le(&mut w, input.sequence)?;
+            }
+            Ok(w.finish())
+        }
+    }
+
+    /// Calculates `hash_outputs` according to BIP143 semantics.`
+    ///
+    /// For BIP143 (Witness and Compatibility sighash) documentation, see here:
+    ///
+    /// - https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki
+    ///
+    /// TODO: memoize
+    fn hash_outputs(&self, index: usize, sighash_flag: Sighash) -> TxResult<Blake2b256Digest> {
+        match sighash_flag {
+            Sighash::All | Sighash::AllACP | Sighash::AllNoInput | Sighash::AllNoInputACP => {
+                let mut w = Blake2b256Writer::default();
+                for output in self.vout.iter() {
+                    output.write_to(&mut w)?;
+                }
+                Ok(w.finish())
+            }
+            Sighash::Single | Sighash::SingleACP | Sighash::SingleNoInput | Sighash::SingleNoInputACP => {
+                if index < self.vout.len() {
+                    let mut w = Blake2b256Writer::default();
+                    self.vout[index].write_to(&mut w)?;
+                    Ok(w.finish())
+                } else {
+                    Ok(Blake2b256Digest::default())
+                }
+            }
+            Sighash::SingleReverse | Sighash::SingleReverseACP | Sighash::SingleReverseNoInput | Sighash::SingleReverseNoInputACP => {
+                if index < self.vout.len() {
+                    let mut w = Blake2b256Writer::default();
+                    self.vout[self.vout.len() - 1 - index].write_to(&mut w)?;
+                    Ok(w.finish())
+                } else {
+                    Ok(Blake2b256Digest::default())
+                }
+            }
+            _ => Ok(Blake2b256Digest::default()),
+        }
+    }
+}
+
+impl Transaction for HandshakeTx {
+    type Digest = Blake2b256Digest;
+    type TxError = TxError;
+    type TXID = TXID;
+    type TxOut = TxOut;
+    type TxIn = HandshakeTxIn;
+    type HashWriter = Blake2b256Writer;
+    type SighashArgs = SighashArgs;
+
+
+    /// Instantiate a new `BitcoinTx`. This always returns a `BitcoinTx::Legacy`
+    fn new<I, O>(version: u32, vin: I, vout: O, locktime: u32) -> Self
```

Comment:
Maybe I should `impl WitnessTransaction` instead of `Transaction`, the `new` method doesn't accept witnesses here